x-pack/plugin/ml/src/main/java/org/elasticsearch/xpack/ml/autoscaling/NodeAvailabilityZoneMapper.java

```diff
@@ -0,0 +1,162 @@
+     *         are not configured then the map will contain one entry mapping an empty list to a collection of all nodes. If
+     *         it is too early in the lifecycle of the node to know the answer then an empty map will be returned. An empty
+     *         map will also be returned if there are no ML nodes in the cluster. (These two empty map return scenarios can be
+     *         distinguished by calling one of the other methods.)
+     */
+    public Map<List<String>, Collection<DiscoveryNode>> getMlNodesByAvailabilityZone() {
+        return mlNodesByAvailabilityZone;
+    }
+
+    /**
+     * @return The number of availability zones in the cluster. If availability zones are not configured this will be 1.
+     *         If it is too early in the lifecycle of the node to know the answer then {@link OptionalInt#empty()} will be returned.
+     *         Unlike {@link #getNumAvailabilityZones()}, it is possible this method will return 0, as it is possible
+     *         for a cluster to have no ML nodes.
+     */
+    public OptionalInt getNumMlAvailabilityZones() {
+        return (lastDiscoveryNodes == null) ? OptionalInt.empty() : OptionalInt.of(mlNodesByAvailabilityZone.size());
+    }
+
+    @Override
+    public synchronized void clusterChanged(ClusterChangedEvent event) {
+        if (lastDiscoveryNodes == null || event.nodesChanged()) {
+            lastDiscoveryNodes = event.state().nodes();
+            buildNodesByAvailabilityZone();
+        }
+    }
+
+    private synchronized void buildNodesByAvailabilityZone() {
+        if (lastDiscoveryNodes == null) {
+            allNodesByAvailabilityZone = Map.of();
+            mlNodesByAvailabilityZone = allNodesByAvailabilityZone;
+            return;
+        }
+
+        Collection<DiscoveryNode> nodes = lastDiscoveryNodes.getNodes().values();
+
+        if (awarenessAttributes.isEmpty()) {
+            allNodesByAvailabilityZone = Map.of(List.of(), nodes);
+            mlNodesByAvailabilityZone = Map.of(
+                List.of(),
+                nodes.stream().filter(n -> n.getRoles().contains(DiscoveryNodeRole.ML_ROLE)).toList()
+            );
+            return;
+        }
+
+        Map<List<String>, Collection<DiscoveryNode>> updatedNodesByAvailabilityZone = new HashMap<>();
+        Map<List<String>, Collection<DiscoveryNode>> updatedMlNodesByAvailabilityZone = new HashMap<>();
+        for (DiscoveryNode node : nodes) {
+            List<String> orderedNodeAttributeValues = awarenessAttributes.stream()
+                .map(a -> node.getAttributes().get(a))
+                .filter(Objects::nonNull)
+                .toList();
+            if (orderedNodeAttributeValues.size() != awarenessAttributes.size()) {
+                // This indicates bad configuration - there shouldn't be nodes that don't have all the awareness attributes
+                logger.trace(
```

Comment:
Should this be a `warn` instead? It sounds like a serious enough misconfiguration that the user should find out.

Reply:
It won't ever happen in Elastic Cloud.

It could easily happen in a self-managed cluster. If it's a warning it could spam the log continually.

The way it gets reported for shard allocation is that it's silently treated as a "NO" unless debug is enabled, in which case a debug message gets generated: https://github.com/elastic/elasticsearch/blob/788df35e339d5f1c90592d5e12c929b3f959bc78/server/src/main/java/org/elasticsearch/cluster/routing/allocation/decider/AwarenessAllocationDecider.java#L174

So maybe the equivalent here is to make this a `debug` log?

Reply:
Fair enough. Yeah, `debug` should be good enough.

Reply:
Done in c57a8b249e5bfd70b0cb1a14941449a3ded55215